Xamarin.Forms.Platform.Android/CollectionView/EmptyViewAdapter.cs

```diff
@@ -62,50 +112,162 @@ public override void OnViewRecycled(Object holder)
 
 		public override void OnBindViewHolder(RecyclerView.ViewHolder holder, int position)
 		{
-			if (EmptyView == null)
+			if (IsHeader(position))
 			{
+				if (holder is TemplatedItemViewHolder templatedItemViewHolder)
+				{
+					BindTemplatedItemViewHolder(templatedItemViewHolder, Header);
+				}
+
 				return;
 			}
 
-			if (holder is SimpleViewHolder emptyViewHolder && emptyViewHolder.View != null)
+			if (IsFooter(position))
 			{
-				// For templated empty views, this will happen on bind. But if we just have a plain-old View,
-				// we need to add it as a "child" of the ItemsView here so that stuff like Visual and FlowDirection
-				// propagate to the controls in the EmptyView
-				ItemsView.AddLogicalChild(emptyViewHolder.View);
+				if (holder is TemplatedItemViewHolder templatedItemViewHolder)
+				{
+					BindTemplatedItemViewHolder(templatedItemViewHolder, Footer);
+				}
+
+				return;
 			}
-			else if (holder is TemplatedItemViewHolder templatedItemViewHolder && EmptyViewTemplate != null)
+
+			if (IsEmpty(position))
 			{
-				// Use EmptyView as the binding context for the template
-				templatedItemViewHolder.Bind(EmptyView, ItemsView);
+				if (holder is SimpleViewHolder emptyViewHolder && emptyViewHolder.View != null)
+				{
+					// For templated empty views, this will happen on bind. But if we just have a plain-old View,
+					// we need to add it as a "child" of the ItemsView here so that stuff like Visual and FlowDirection
+					// propagate to the controls in the EmptyView
+					ItemsView.AddLogicalChild(emptyViewHolder.View);
+				}
+				else if (holder is TemplatedItemViewHolder templatedItemViewHolder && EmptyViewTemplate != null)
+				{
+					// Use EmptyView as the binding context for the template
+					templatedItemViewHolder.Bind(EmptyView, ItemsView);
+				}
 			}
 		}
 
 		public override RecyclerView.ViewHolder OnCreateViewHolder(ViewGroup parent, int viewType)
 		{
 			var context = parent.Context;
 
-			var template = EmptyViewTemplate;
+			if (viewType == _headerViewType)
```

Comment:
A switch statement might be nicer here.

Reply:
Yes, a switch is nicer in in this case but the viewtype used to identify header, footer, etc. are not constant (to force OnCreateViewHolder to run again in case neccesary). That's why we use If's in this case.